java/arcs/sdk/Particle.kt

```diff
@@ -24,9 +24,42 @@ typealias Particle = arcs.core.host.Particle
  * Base class used by `schema2kotlin` code-generator tool to generate a class containing all
  * declared handles.
  */
-abstract class HandleHolderBase(
-    override val handles: MutableMap<String, Handle>,
-    override val entitySpecs: Map<String, EntitySpec<out Entity>>
+open class HandleHolderBase(
+    private val particleName: String,
+    private val entitySpecs: Map<String, EntitySpec<out Entity>>
 ) : HandleHolder {
+    val handles = mutableMapOf<String, Handle>().withDefault { handleName ->
+        checkHandleIsValid(handleName)
+        throw NoSuchElementException(
+            "Handle $handleName has not been initialized in $particleName yet."
+        )
+    }
+
+    override fun getEntitySpec(handleName: String): EntitySpec<out Entity> {
+        checkHandleIsValid(handleName)
+        return entitySpecs.getValue(handleName)
+    }
+
+    override fun getHandle(handleName: String): Handle {
+        checkHandleIsValid(handleName)
+        return handles.getValue(handleName)
+    }
+
+    override fun setHandle(handleName: String, handle: Handle) {
+        checkHandleIsValid(handleName)
+        require(!handles.containsKey(handleName)) {
+            "$particleName.$handleName has already been initialized."
+        }
+        handles[handleName] = handle
+    }
+
     override fun clear() = handles.clear()
+
+    private fun checkHandleIsValid(handleName: String) {
+        if (!entitySpecs.containsKey(handleName)) {
+            throw NoSuchElementException(
+                "Particle $particleName does not have a handle with name $handleName."
```

Comment:
does not have an 'EntitySpec with name $handleName' ?

Reply:
This one was a little subtle. The entire `List<EntitySpec>` gets passed in when this object is created, so we can expect that it has a full list of handle names and is reasonably authoritative. Whereas the handle list can grow over time as handles get created, so we want to check handle names against the entitySpecs map to see if they are valid.

I'll add a comment.